pkg/util/helper/taint_test.go

```diff
@@ -96,38 +86,26 @@ func TestUpdateClusterControllerTaint(t *testing.T) {
 				taintsToRemove: []*corev1.Taint{unreachableTaintTemplate.DeepCopy()},
 			},
 			wantTaints: []corev1.Taint{*notReadyTaintTemplate},
-			wantErr:    false,
 		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			ctx, cancel := context.WithCancel(context.Background())
-			defer cancel()
-
 			cluster := &clusterv1alpha1.Cluster{
 				ObjectMeta: metav1.ObjectMeta{Name: "member"},
 				Spec: clusterv1alpha1.ClusterSpec{
 					Taints: tt.args.taints,
 				},
 			}
-			c := fakeclient.NewClientBuilder().WithScheme(gclient.NewSchema()).WithObjects(cluster).Build()
-
-			if err := UpdateClusterControllerTaint(ctx, c, tt.args.taintsToAdd, tt.args.taintsToRemove, cluster); (err != nil) != tt.wantErr {
-				t.Errorf("UpdateClusterControllerTaint() error = %v, wantErr %v", err, tt.wantErr)
-			}
-
-			if err := c.Get(ctx, client.ObjectKey{Name: cluster.Name}, cluster); err != nil {
-				t.Fatalf("Failed to get cluster %s: %v", cluster.Name, err)
-			}
 
-			if len(cluster.Spec.Taints) != len(tt.wantTaints) {
-				t.Errorf("Cluster gotTaints = %v, want %v", cluster.Spec.Taints, tt.wantTaints)
+			taints := SetCurrentClusterTaints(tt.args.taintsToAdd, tt.args.taintsToRemove, cluster)
+			if len(taints) != len(tt.wantTaints) {
+				t.Errorf("Cluster gotTaints = %v, want %v", taints, tt.wantTaints)
 			}
-			for i := range cluster.Spec.Taints {
-				if cluster.Spec.Taints[i].Key != tt.wantTaints[i].Key ||
-					cluster.Spec.Taints[i].Value != tt.wantTaints[i].Value ||
-					cluster.Spec.Taints[i].Effect != tt.wantTaints[i].Effect {
-					t.Errorf("Cluster gotTaints = %v, want %v", cluster.Spec.Taints, tt.wantTaints)
+			for i := range taints {
+				if taints[i].Key != tt.wantTaints[i].Key ||
```

Comment:
Can we use `reflect.DeepEqual`?

Reply:
I'm afraid we cannot. `Taints` has a field `TimeAdd` which is the timestamp. We cannot ensure this field is equal.